### ISSUE TITLE
Two layernorms one after the other in actual settings during running

### ISSUE THREAD

**Issue by BestJuly** [microsoft/Swin-Transformer#213]
Hi, @zeliu98. First, thanks for the great work~

I was doing some profiling on SwinTransformer and I found there were two layernorm layer one after the other at the very beginning of the model. Is this a bug or a feature? More details are listed below.

![image](https://user-images.githubusercontent.com/19769279/170160062-fd2cae39-12d8-4f46-8636-2393f8da5431.png)

1. The first layer norm is defined in module [PatchEmbed](https://github.com/microsoft/Swin-Transformer/blob/main/models/swin_transformer.py#L447);
2. Self.ape and self.pos_drop used in the [forwarding pass](https://github.com/microsoft/Swin-Transformer/blob/main/models/swin_transformer.py#L561). However, it is actually inactivated;
3. Then the processing start with BasicLayer module, and the first submodule is SwinTransformerBlock, where the first processing is using [layernorm](https://github.com/microsoft/Swin-Transformer/blob/main/models/swin_transformer.py#L239).

The main concern is here, I use comment in the corresponding part.
```python
def forward_features(self, x):
        x = self.patch_embed(x)  # The last operation is layernorm, which is activated
        if self.ape:  # Actually inactivated
            x = x + self.absolute_pos_embed  # Actually inactivated
        x = self.pos_drop(x)  # Actually inactivated

        for layer in self.layers:
            x = layer(x)  # The first operation is layernorm

        x = self.norm(x)  
        x = self.avgpool(x.transpose(1, 2))  
        x = torch.flatten(x, 1)
        return x
```

There are learnable parameters in layernorm module. Two layernorm layers will cause different results compared to one layernorm layer.

Here is a simple validation code for the layernorm part:
```python
import torch
import torch.nn as nn

def copy_one_tensor(input, requires_grad=True):
    input1 = input.clone().detach().requires_grad_(requires_grad).cuda()
    return input1


class OneLayerNorm(nn.Module):
    def __init__(self, C, H, W):
        super(OneLayerNorm, self).__init__()
        self.ln1 = nn.LayerNorm([C, H, W])
        self.apply(self._init_weights)

    def forward(self, x):
        output = self.ln1(x)
        return output

    def _init_weights(self, m):
        if isinstance(m, nn.LayerNorm):
            nn.init.constant_(m.bias, 0)
            nn.init.constant_(m.weight, 1.0)

class TwoLayerNorm(nn.Module):
    def __init__(self, C, H, W):
        super(TwoLayerNorm, self).__init__()
        self.ln1 = nn.LayerNorm([C, H, W])
        self.ln2 = nn.LayerNorm([C, H, W])
        self.apply(self._init_weights)

    def forward(self, x):
        output = self.ln2(self.ln1(x))
        return output
    
    def _init_weights(self, m):
        if isinstance(m, nn.LayerNorm):
            nn.init.constant_(m.bias, 0)
            nn.init.constant_(m.weight, 1.0)


N, C, H, W = 20, 5, 10, 10

input = torch.randn((N, C, H, W))
d_loss_tensor = torch.randn((N, C, H, W)).cuda()
input1 = copy_one_tensor(input)
input2 = copy_one_tensor(input)


model1 = OneLayerNorm(C, H, W).cuda()
model2 = TwoLayerNorm(C, H, W).cuda()

criterion1 = nn.MSELoss()
criterion2 = nn.MSELoss()

optim1 = torch.optim.Adam(model1.parameters(), lr=0.1)
optim2 = torch.optim.Adam(model2.parameters(), lr=0.1)

for _ in range(10):
    y = model1(input1)
    loss1 = criterion1(d_loss_tensor, y)
    optim1.zero_grad()
    loss1.backward()
    optim1.step()

for _ in range(10):
    y = model2(input2)
    loss2 = criterion2(d_loss_tensor, y)
    optim2.zero_grad()
    loss2.backward()
    optim2.step()


new_x = torch.randn((N, C, H, W))
new_x1 = copy_one_tensor(new_x)
new_x2 = copy_one_tensor(new_x)


with torch.no_grad():
    output1 = model1(new_x1)
    output2 = model2(new_x2)

print(torch.allclose(output1, output2, rtol=1e-05, atol=1e-08))
# the output is False
```

I understand if activating the absolute position embedding and the position dropout layer, it would be reasonable. And currectly, it might not affect the final performance a lot, but I wonder whether this is as expected or not?

**Comment by zeliu98** [comment]
Hi @BestJuly, thank you for your carefully profiling. There is a little difference between LN1 (the layernorm at the end of patch_embed) and LN2 (the layernorm in the first swin block). As can be seen from the following diagram, LN1 is located in the main branch while LN2 is located in the residual branch of the first swin block.
```mermaid
graph LR
patch_embed --> LN1
LN1 --> add((+))
LN1 --> LN2
LN2 --> ...
... --> add((+))
```
As you have mentioned, it maybe ok if we remove LN2 when absolute_pos_embed  and pos_drop is inactivated. However, we don't do this mostly due to the simplicity of coding where the first Swin block needs to be distinguished from other blocks.

Anyway, maybe you could probably try to remove LN2 if it causes too much additional cost for you.

**Comment by BestJuly** [comment]
@zeliu98 Thank you for your answering. As you mentioned, the LN2 is located in the second path, which means for the second path of the first block, the input will go through two LN layers.

I think it may not affect too much if removed one LN layer here. 
If one LN will be removed, it should be LN2 as the result after LN1 will be used as another path for the residual add operation. However, I just removed LN1 for convenience and train for 9k iterations with `batchsize=192` when pre-training on ImageNet-1K, it seems to be no huge differences from the loss curves, for both fp32 and fp16 settings.

As I noticed from SwinV2, for the training stability, additional LN will be added every several blocks. Therefore, though the networks would be slightly different, for models which are not that big, it seems that more LNs or slightly less LNs would not affect the performance too much.
